### PR TITLE
[feature] Move linear solver to GPUs

### PR DESCRIPTION
**Pull request by toxa81**

### CODE REVIEW

File: src/api/sirius_api.cpp [hunk 1]
```diff
@@ -5997,6 +5997,16 @@ void sirius_linear_solver(void* const* handler__, double const* vkq__, int const
             auto Sphi_wf = sirius::wave_function_factory<double>(sctx, kp, wf::num_bands(sctx.num_bands()), wf::num_mag_dims(0), false);
 
             // TODO: allocate all wave-functions on device (if processing_unit == sdk::memory_t::GPU)
+            if (sctx.processing_unit() == sddk::device_t::GPU){
```
Comment (toxa81) [inline src/api/sirius_api.cpp:6000]: `if` statement is not needed here

File: src/api/sirius_api.cpp [hunk 2]
```diff
@@ -5997,6 +5997,16 @@ void sirius_linear_solver(void* const* handler__, double const* vkq__, int const
...
+
+                sddk::memory_t mem = sctx.processing_unit_memory_t();
+
+                auto mg1 = psi_wf->memory_guard(mem, wf::copy_to::device);
+                auto mg2 = dpsi_wf->memory_guard(mem, wf::copy_to::device | wf::copy_to::host);
+                auto mg3 = dvpsi_wf->memory_guard(mem, wf::copy_to::device);
+                auto mg4 = tmp_wf->memory_guard(mem, wf::copy_to::device);
```
Comment (toxa81) [inline src/api/sirius_api.cpp:6007]: for temporaty, no copy is needed

File: src/api/sirius_api.cpp [hunk 3]
```diff
@@ -5996,6 +5996,24 @@ void sirius_linear_solver(void* const* handler__, double const* vkq__, int const
             auto Hphi_wf = sirius::wave_function_factory<double>(sctx, kp, wf::num_bands(sctx.num_bands()), wf::num_mag_dims(0), false);
             auto Sphi_wf = sirius::wave_function_factory<double>(sctx, kp, wf::num_bands(sctx.num_bands()), wf::num_mag_dims(0), false);
 
+            sddk::memory_t mem = sctx.processing_unit_memory_t(); // "auto mem" should do as well
+
+            // TODO: allocate all wave-functions on device (if processing_unit == sdk::memory_t::GPU)
+//          if (sctx.processing_unit() == sddk::device_t::GPU){
+            { // START scope of memory guards
+
```
Comment (toxa81) [inline src/api/sirius_api.cpp:6004]: It is better to use std::vector<memory_guard> and push to it

File: src/api/sirius_api.cpp [hunk 4]
```diff
@@ -6035,6 +6054,7 @@ void sirius_linear_solver(void* const* handler__, double const* vkq__, int const
                 100, // iters
                 1e-13 // tol
             );
+            }  // END scope of memory guards
```
Comment (toxa81) [inline src/api/sirius_api.cpp:6057]: here the vector of mem. guards can be cleaned

File: src/api/sirius_api.cpp (hunk 3 above)
Comment (gsavva) [reply src/api/sirius_api.cpp:6004]: done

File: src/api/sirius_api.cpp (hunk 4 above)
Comment (gsavva) [reply src/api/sirius_api.cpp]: done

**Pull request closed without merging**